### PR TITLE
add CTRL Model

### PR DESCRIPTION
**Pull request by JunnYu**
飞桨论文复现挑战赛（第四期）CTRL: A Conditional Transformer Language Model for Controllable Generation 论文复现提交。

### CODE REVIEW

**Comment by yingyibiao** [comment]
前向输出还需要再定位清楚不一致的op！

**Comment by ZeyuChen** [comment]
这个PR还有效吗？@yingyibiao

**Comment by yingyibiao** [comment]
> 这个PR还有效吗？@yingyibiao

有效

File: paddlenlp/transformers/ctrl/modeling.py [hunk 1]
```diff
@@ -0,0 +1,764 @@
...
+        ffn_output = self.ffn(out2)
+        ffn_output = self.dropout2(ffn_output)
+        out2 = out1 + ffn_output
+
+        outputs = (out2, ) + attn_outputs[1:]
+        return outputs
+
+
+class CTRLPreTrainedModel(PretrainedModel):
+    base_model_prefix = "transformer"
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/modeling.py]: The `base_model_prefix` for CTRL should be 'ctrl'

File: paddlenlp/transformers/ctrl/modeling.py [hunk 2]
```diff
@@ -0,0 +1,764 @@
...
+            if layer._padding_idx is not None:
+                emb_weight = layer.weight.numpy()
+                emb_weight[layer._padding_idx] = np.zeros_like(emb_weight[
+                    layer._padding_idx])
+                layer.weight.set_value(paddle.to_tensor(emb_weight))
+        elif isinstance(layer, nn.LayerNorm):
+            layer.weight.set_value(paddle.ones_like(layer.weight))
+            layer.bias.set_value(paddle.zeros_like(layer.bias))
+
+    def greedy_search(
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/modeling.py]: Why put decode strategy here?

File: paddlenlp/transformers/ctrl/modeling.py [hunk 3]
```diff
@@ -0,0 +1,764 @@
...
+            model_kwargs["past"] = None
+
+        return model_kwargs
+
+
+@register_base_model
+class CTRLModel(CTRLPreTrainedModel):
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.pd_config = pd_config = Config(**kwargs)
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/modeling.py]: Please use paddlenlp-style config. See other models for details

File: paddlenlp/transformers/ctrl/modeling.py [hunk 4]
```diff
@@ -0,0 +1,764 @@
...
+    output = paddle.matmul(attention_weights, v)
+
+    return output, attention_weights
+
+
+class MultiHeadAttention(nn.Layer):
+    def __init__(self, d_model_size, num_heads):
+        super().__init__()
+        self.num_heads = num_heads
+        self.d_model_size = d_model_size
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/modeling.py]: Better to use hidden_size instead of d_model_size

File: paddlenlp/transformers/ctrl/modeling.py [hunk 5]
```diff
@@ -0,0 +1,764 @@
...
+            model_kwargs["past"] = outputs.past_buckets_states
+        else:
+            model_kwargs["past"] = None
+
+        return model_kwargs
+
+
+@register_base_model
+class CTRLModel(CTRLPreTrainedModel):
+    def __init__(self, **kwargs):
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/modeling.py]: Please add detailed documentation for your model. See bert for example

File: paddlenlp/transformers/ctrl/modeling.py [hunk 6]
```diff
@@ -0,0 +1,764 @@
...
+import paddle.nn as nn
+import paddle.nn.functional as F
+from paddle.nn import CrossEntropyLoss, MSELoss
+from .. import PretrainedModel, register_base_model
+
+from .utils import (
+    BaseModelOutputWithPast,
+    CausalLMOutputWithPast,
+    Config,
+    SequenceClassifierOutput, )
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/modeling.py]: Please don't use such output format. Use PaddleNLP-style output. See other model for example

File: paddlenlp/transformers/ctrl/tokenizer.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+    def _tokenize(self, text):
+        split_tokens = []
+        re = try_import("regex")
+        words = re.findall(r"\S+\n?", text)
+        for token in words:
+            split_tokens.extend([t for t in self.bpe(token).split(" ")])
+        return split_tokens
+
+    def _convert_token_to_id(self, token):
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/tokenizer.py:250]: Please add more doc for function like this.

**Review by smallv0221 — CHANGES_REQUESTED** [review]
Things in utils.py is too familiar to HuggingFace. Please change them to PaddleNLP-style. And add more doc. You can see other model for details.

File: paddlenlp/transformers/ctrl/tokenizer.py (hunk 7 above)
Comment (JunnYu) [reply paddlenlp/transformers/ctrl/tokenizer.py:250]: done

File: paddlenlp/transformers/ctrl/modeling.py (hunk 6 above)
Comment (JunnYu) [reply paddlenlp/transformers/ctrl/modeling.py]: done

File: paddlenlp/transformers/ctrl/modeling.py (hunk 5 above)
Comment (JunnYu) [reply paddlenlp/transformers/ctrl/modeling.py]: done

File: paddlenlp/transformers/ctrl/modeling.py (hunk 4 above)
Comment (JunnYu) [reply paddlenlp/transformers/ctrl/modeling.py]: done

File: paddlenlp/transformers/ctrl/modeling.py (hunk 3 above)
Comment (JunnYu) [reply paddlenlp/transformers/ctrl/modeling.py]: done

File: paddlenlp/transformers/ctrl/modeling.py (hunk 1 above)
Comment (JunnYu) [reply paddlenlp/transformers/ctrl/modeling.py]: done

**Comment by JunnYu** [comment]
CTRL权重文件
链接：https://pan.baidu.com/s/1JsxGtNBHK7VSlUnr_mByCA 
提取码：4834 
--来自百度网盘超级会员V5的分享

File: paddlenlp/transformers/ctrl/modeling.py [hunk 8]
```diff
@@ -0,0 +1,844 @@
...
+        r'''
+        The CTRLModel forward method, overrides the `__call__()` special method.
+
+        Args:
+            input_ids (Tensor):
+                Indices of input sequence tokens in the vocabulary. They are
+                numerical representations of tokens that build the input sequence.
+                Its data type should be `int64` and it has a shape of [batch_size, sequence_length].
+            cache (Tuple[Tuple[Tensor]], optional):
+                Contains pre-computed hidden-states (key and values in the attention blocks) as computed by the model. Can be used to speed up sequential decoding. The `input_ids` which have their past given to this model should not be passed as input ids as they have already been computed.
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/modeling.py]: 这里的文档单行不要太长，最好每85个字符左右换行一次

File: paddlenlp/transformers/ctrl/__init__.py [hunk 9]
```diff
@@ -0,0 +1,2 @@
+from .modeling import *
```
Comment (smallv0221) [inline paddlenlp/transformers/ctrl/__init__.py]: This file could be empty.

File: paddlenlp/transformers/ctrl/modeling.py (hunk 8 above)
Comment (JunnYu) [reply paddlenlp/transformers/ctrl/modeling.py]: done

File: paddlenlp/transformers/ctrl/__init__.py (hunk 9 above)
Comment (JunnYu) [reply paddlenlp/transformers/ctrl/__init__.py]: done

**Review by smallv0221 — DISMISSED** [review]
LGTM

**Comment by yingyibiao** [comment]
需要在example/text_generation下面新增一个使用案例

**Comment by JunnYu** [comment]
> 需要在example/text_generation下面新增一个使用案例

已添加，测试生成结果如下。
image

**Review by yingyibiao — APPROVED** [review]

**Pull request merged**